# Murder in Sequel City: Can you solve the mystery?

![](data/images/spy.png)

## Goals

This case will introduce you to the basics of databases and give you a first exposure to the SQL programming language, which we use to communicate with them.

## Introduction

It is a misty September morning in Sequel City, California. You are relaxing at your office, reading the newspaper and drinking coffee, when the telephone rings. It's the familiar voice of one of your high-profile contacts from Washington. He wants you, the most renowned private investigator in town, to help the FBI solve a top-secret cold case. Will you be up to the challenge?

![The Red Telephone](data/images/phone.jpg)

He tells you there was a murder back in 2018, but the first investigation didn't produce any definitive results. The bad news is that the original crime scene data was lost because it was stored in Excel workbooks, and the Excel files got corrupted some time after the case was archived. As if that was not enough, the detective who led the inquiry back then forgot many of the key details of the case! The only thing she vaguely recalls is that the crime occurred sometime on Feb. 18, 2018. The good news is that the data was copied to a "relational database" (more on that later) before the Excel files were corrupted. All the clues that you will need to solve the case are buried there.

You are always down for a good mystery, so you *of course* say yes!

## The database enigma

Many people know how to handle data in Excel spreadsheets, which are intuitive and easy to use and access. But that is *so mainstream*! There are other tools with greater potential.

![from_excel_to_db](./data/images/from_excel_to_db.png)

Although Excel is great, it has its limitations. For starters, [Excel](https://support.microsoft.com/en-us/office/excel-specifications-and-limits-1672b34d-7043-467e-8e27-269d656771c3) files have a maximum size of 1,048,576 rows by 16,384 columns, and in practice the limit is much smaller than that because many personal computers can't handle that amount of data along with Excel's user interface at the same time. Furthermore, while having a single very large table may not be so bad, the situation becomes more complicated when you have to look up the data in that table in another very large table, which in turn references another table, etc. After a bunch of [`VLOOKUP`](https://support.microsoft.com/en-us/office/vlookup-function-0bbc8083-26fe-4963-8ab8-93a18ad188a1)s across several large tables, not only will your Excel crash, but you will also risk losing your data!


On top of that, collaboration is difficult in Excel - how would you allow your colleagues to work together on the same spreadsheet as you? One option could be to use Excel Online or Google Spreadsheets, but although these tools are excellent for collaboration on small datasets, they are not optimal for bigger tasks like the one we are describing.

For these reasons, and many more, professional data analysts (and professional private detectives too!) use [**relational databases**](https://en.wikipedia.org/wiki/Relational_database) instead. Relational databases are purpose-built to do all the tasks we have just mentioned in a very efficient and secure way.

### What is a relational database?

There's no single definition for the word database. In general, databases are systems for managing information. Databases can have varying amounts of structure imposed on the data. When the data is more structured, it can help people and computers work with the data more efficiently.

Relational databases are probably the most well-known kind of database. At their heart, relational databases are made up of **tables**, which are a lot like spreadsheets. Each column in the table has a name and a data type (text, number, etc.), and each row in the table is a specific instance of whatever the table is about. Relational databases get their name from the term **relations**, which is another way to refer to tables. So you could say that relational databases are indeed "table-based databases"!

For example, one of the tables (relations) in the cold case database contains information about people. The first few rows of that table are below:


| id |                name |  driver_id |               address |         ssn |    income |
|---:|--------------------:|-----------:|----------------------:|------------:|----------:|
|  1 |         Herve Lilie | 7023162704 | 9663 Mockingbird Lane | 156-14-0067 | \$63163.25 |
|  2 |    Jacqueline Colam | 4722687668 |  1396 Westport Circle | 733-53-7490 | \$94851.22 |
|  3 |       Keenan Treker | 9198946080 |   4 Glendale Junction | 380-51-1574 | \$58070.55 |
|  4 |     Kimbell Everill |   39201465 |      107 Hovde Circle | 415-98-5200 | \$40770.06 |
|  5 | Abramo De Bernardis | 7411450154 |  8251 Maple Wood Park | 657-03-1688 | \$89059.97 |

You can see that this table has an `id` column to uniquely identify each person. We call identifying columns **primary keys**. Thus, `id` is this table's primary key. You can also see other columns like `name`, `driver_id`, `address`, etc. These are called **attributes**. Each row corresponds to one person and is called a **tuple**.

## The conspiracy puzzle

As a seasoned private detective, the first thing you do when you start on a new case is to rush to buy some pins and thread at Walmart and make an evidence board, [*don't you?*](https://qr.ae/pvOGXe) Evidence boards are great because they help you see all the evidence at a glance.

![investigation](data/images/investigation.jpg)


Turns out we can have evidence boards in relational databases too! We call them **Entity Relationship Diagrams**, or **ERD**s for short.

An ERD is a visual representation of the relationships between all the relevant tables within a database. You can find the ERD of our cold case database below. Take some time to go over all the tables and understand the structure of the database. The diagram shows that each table has a name (at the top) and a list of the attributes with their data types. A key icon indicates that the column is the primary key of the corresponding table. The blue lines mean that two tables are linked to each other. We link tables using **foreign keys**. A foreign key is an attribute that is used to reference data in one table from another table. For example, the attribute `driver_id` in the `individual` table is a foreign key because it points to an attribute in another table, the `id` attribute of the `drivers` table.

![ERD](data/images/er.png)

By knowing how all the relations in the database are linked to each other, you can narrow down your research to find out exactly what data you need to extract and where to look for it.

## The query inquiry

Just as a detective interrogates a suspect, a data analyst can interrogate a database. We call that action **querying** the database.

Databases are very particular subjects that won't cooperate if you don't ask the right way. As rightful citizens of Sequel City (databases enjoy the same civil rights that human beings have here), they demand their interrogations to be conducted in the language of the land. That language is **SQL** (pronounced "sequel," just like the city) and stands for **Structured Query Language**.

Are you ready to grill the database? Let's make it speak!

### Initializing the database

Let's load our database. We will be using a database that is in the [SQLite](https://www.sqlite.org/index.html) file format. This format is very popular for small databases. In fact, chances are your cell phone uses hundreds of SQLite databases for stuff like keeping your contact lists and apps up-to-date.

In order to load the SQLite file (which you can find at [`data/crime_database.db`](data/crime_database.db)), we need to import the `sqlalchemy` Python library and run some boilerplate code (below). Don't worry about learning that code, it's not SQL.

In [1]:
%FETCH https://workspace-amz-da6.c1-learning.com/case.sql_basics_fellow/data/crime_database.db crime_database

Start downloading from URL https://workspace-amz-da6.c1-learning.com/case.sql_basics_fellow/data/crime_database.db
Finished downloading 1384448 bytes from URL https://workspace-amz-da6.c1-learning.com/case.sql_basics_fellow/data/crime_database.db
Writing downloaded data to file crime_database
Finished writing file


In [2]:
%LOAD crime_database RW

### The basics of queries

If you were to look at the data in this database, you would see that the tables are huge! There are so many data points; it is simply not possible to go through the tables row by row to find the information we need. What are we supposed to do?

This is where queries come in. A good thing about SQL is that queries read like natural English (for the most part), although they have a much more limited vocabulary. The word that databases like the most is **`SELECT`**, which lets you access data from a table or set of tables. They like it even more if once you have finished your query, you end it with a semicolon (**`;`**), which lets it know you're done asking. Saying `SELECT` and `;` to a database is like saying "please" and "thank you" to a Sesame Street character. It makes it sing!

### Example 1

To ask a database to show you the data from all the columns in a table, you use the **`*`** character. To specify the table, you use the **`FROM`** keyword. For example, below we query all the data from all the columns of the `interrogation` table (the cell starts with `%%sql`, which is a piece of code we add to be able to execute SQL code inside a notebook, but it isn't actual SQL language syntax):

In [3]:
SELECT * FROM interrogation;

+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------+
| description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | individual_id |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------+
| The intense interest aroused in the public by what was known at the time as “The Styles Case” has now somewhat subsided. Nevertheless, in view of the world-wide notoriety which attended it, I have been asked, both by my friend Poirot and the family themselves, to write an account of the whole story. This, we trust, will effectually silence the sensational rumours which still persist.                                                                                                                                                      

Locate the `interrogation` table in the Entity-Relationship Diagram. Are all column names the same?

If you find that the output is taking too long to appear, you might want to limit the number of retrieved tuples (rows). You may want to have only 10 results, for instance. For this, you would use, rather unsurprisingly, the **`LIMIT`** keyword:

In [ ]:
SELECT * FROM interrogation LIMIT 10;

### Exercise 1

It's your turn. Ask the database (and remember to do it politely) to show you all the data from the `individual` table. Limit your results to only 20 rows.

In [2]:
%FETCH https://workspace-amz-da6.c1-learning.com/case.sql_basics_fellow/data/crime_database.db crime_database


Start downloading from URL https://workspace-amz-da6.c1-learning.com/case.sql_basics_fellow/data/crime_database.db
Finished downloading 1384448 bytes from URL https://workspace-amz-da6.c1-learning.com/case.sql_basics_fellow/data/crime_database.db
Writing downloaded data to file crime_database
Finished writing file


In [8]:
%LOAD crime_database RW


In [5]:
SELECT * FROM individual LIMIT 20;

+----+---------------------+------------+------------------------+-------------+----------+
| id | name                | driver_id  | address                | ssn         | income   |
+----+---------------------+------------+------------------------+-------------+----------+
| 1  | Herve Lilie         | 7023162704 | 9663 Mockingbird Lane  | 156-14-0067 | 63163.25 |
+----+---------------------+------------+------------------------+-------------+----------+
| 2  | Jacqueline Colam    | 4722687668 | 1396 Westport Circle   | 733-53-7490 | 94851.22 |
+----+---------------------+------------+------------------------+-------------+----------+
| 3  | Keenan Treker       | 9198946080 | 4 Glendale Junction    | 380-51-1574 | 58070.55 |
+----+---------------------+------------+------------------------+-------------+----------+
| 4  | Kimbell Everill     | 39201465   | 107 Hovde Circle       | 415-98-5200 | 40770.06 |
+----+---------------------+------------+------------------------+-------------+----------+
| 5  | Abramo De Bernardis | 7411450154 | 8251 Maple Wood Park   | 657-03-1688 | 89059.97 |
+----+---------------------+------------+------------------------+-------------+----------+
| 6  | Iggy Duferie        | 9434499607 | 447 Sheridan Plaza     | 814-41-8602 | 82433.22 |
+----+---------------------+------------+------------------------+-------------+----------+
| 7  | Helli Bartoleyn     | 5345811245 | 60 Dorton Road         | 310-24-3441 | 43438.83 |
+----+---------------------+------------+------------------------+-------------+----------+
| 8  | Lacy Havik          | 2901087841 | 040 Charing Cross Pass | 514-07-9654 | 70112.23 |
+----+---------------------+------------+------------------------+-------------+----------+
| 9  | Winn Weldon         | 1518171052 | 200 7th Park           | 222-76-2206 | 43550.86 |
+----+---------------------+------------+------------------------+-------------+----------+
| 10 | Marlena Heggison    | 2949944418 | 35 Darwin Avenue       | 417-24-3059 | 46694.42 |
+----+---------------------+------------+------------------------+-------------+----------+
| 11 | Lora Gatrell        | 9591736789 | 63 Golden Leaf Trail   | 304-84-0146 | 62468.31 |
+----+---------------------+------------+------------------------+-------------+----------+
| 12 | Leicester Poag      | 4409075675 | 26 Michigan Point      | 529-37-7536 | 84163.44 |
+----+---------------------+------------+------------------------+-------------+----------+
| 13 | Alverta Cracker     | 1617632252 | 7 Steensland Avenue    | 239-63-0360 | 78203.95 |
+----+---------------------+------------+------------------------+-------------+----------+
| 14 | Lindsy Sheryn       | 9290370920 | 263 Coolidge Plaza     | 240-19-6212 | 32783.11 |
+----+---------------------+------------+------------------------+-------------+----------+
| 15 | Willey Marwood      | 1052448143 | 4139 Schiller Avenue   | 491-11-0921 | 95935.15 |
+----+---------------------+------------+------------------------+-------------+----------+
| 16 | Brian Starking      | 8632303333 | 47509 Jana Court       | 790-98-2600 | 67685.19 |
+----+---------------------+------------+------------------------+-------------+----------+
| 17 | Peggie Copsey       | 3625549700 | 04052 Waubesa Crossing | 306-68-2633 | 84879.49 |
+----+---------------------+------------+------------------------+-------------+----------+
| 18 | Christa Gillatt     | 2521723878 | 8 Ohio Avenue          | 492-67-3180 | 50534.99 |
+----+---------------------+------------+------------------------+-------------+----------+
| 19 | Mariann Row         | 6389055021 | 9 Monica Center        | 468-79-3672 | 87326.58 |
+----+---------------------+------------+------------------------+-------------+----------+
| 20 | Gasper Callender    | 9889303019 | 8940 Novick Point      | 298-38-4233 | 42109.68 |
+----+---------------------+------------+------------------------+-------------+----------+

### Example 2

Let's now say we only want to know about a specific person, not about all the people in the database. Is there a way to filter the table to include only the rows that meet a certain condition, say, a name match?

Fret not! We need only use the **`WHERE`** keyword. For example, the below code asks the database to

1. Locate the `individual` table (with `FROM`),
1. `SELECT` all columns (`*`) from that table,
1. and filter the results to show only those where the `name` column matches the name 'Finlay Mathiassen' (notice the single quotes, they are important to let the database know that the text between them must be treated as ordinary text and not as SQL words).

In [6]:
SELECT * FROM individual WHERE name = 'Finlay Mathiassen';

+-----+-------------------+-----------+-----------------------+-------------+----------+
| id  | name              | driver_id | address               | ssn         | income   |
+-----+-------------------+-----------+-----------------------+-------------+----------+
| 619 | Finlay Mathiassen | 859708861 | 1500 Transport Center | 116-09-4940 | 73479.19 |
+-----+-------------------+-----------+-----------------------+-------------+----------+

### Exercise 2

Following the model from the example, find everything you can about Julianne McClaurie.

In [7]:
SELECT * FROM individual WHERE name = 'Julianne McClaurie';

+-----+--------------------+------------+-----------------+-------------+----------+
| id  | name               | driver_id  | address         | ssn         | income   |
+-----+--------------------+------------+-----------------+-------------+----------+
| 601 | Julianne McClaurie | 7538214119 | 7 Westport Park | 676-70-2756 | 36196.12 |
+-----+--------------------+------------+-----------------+-------------+----------+

### Example 3

Now that we're building rapport with the database (the good cop strategy is really working so far!), we can start digging in a bit more and ask for more nuanced information. Enter the **`AND`** keyword. It lets you include more than one condition in your query. For example, the below query retrieves all the rows in the `crime_scene` table that are robberies <u>and</u> happened in Chicago:

In [8]:
SELECT * FROM crime_scene  
WHERE type = 'robbery' AND city = 'Chicago';

+------------+---------+---------+---------------+---------------------------------------------------------------------------------+
| date       | type    | city    | country       | description                                                                     |
+------------+---------+---------+---------------+---------------------------------------------------------------------------------+
| 2018-11-02 | robbery | Chicago | United States | “I’ve got an old uncle who’s more or less rolling, but he’s no good.” |
+------------+---------+---------+---------------+---------------------------------------------------------------------------------+

### Exercise 3

Let's now investigate the clue given by the original detective. She remembers that the crime was a *murder* that occurred sometime on *Feb 18, 2018* and took place in *Sequel City*.

Get all the rows in the `crime_scene` whose

* `type` is `murder` and
* whose `city` is `Sequel City`.

In [9]:
SELECT * FROM crime_scene  
WHERE type = 'murder' AND city = 'Sequel City';

+------------+--------+-------------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| date       | type   | city        | country       | description                                                                                                                                                                                                                                           |
+------------+--------+-------------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 2018-11-02 | murder | Sequel City | United States | Tuppence turned sharply, but the words hovering on the tip of her tongue remained unspoken, for the man’s appearance and manner did not bear out her first and most natural assumption. She hesitated.                                              |
+------------+--------+-------------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 2018-02-18 | murder | Sequel City | United States | A security camera in the neighborhood captured two figures who may be witnesses. The first one lives at 513 Hamilton Dr. The second one was tentatively identified by one of the police officers as a man named George, who lives on Independence St. |
+------------+--------+-------------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

These are interesting results indeed! Jot down the output of this query. We will likely find it useful down the road.

## The similarity conundrum

Sometimes you only know part of the information you need. SQL can handle that. Special symbols that represent unknown characters are called "wildcards". The most common is the `%` wildcard.

When you place a `%` wildcard in a query string, the SQL system will return results that match the rest of the string exactly, and have anything (or nothing) where the wildcard is. For example, `Ca%a` matches `Canada` and `California` because both words start with "Ca" and end with an "a".

The other, less commonly used wildcard, is `_`. This one means 'match the rest of the text, as long as there's exactly one character in exactly the position of the `_`, no matter what it is. So, `B_b` would match `Bob` and `Bub` but not `Babe` or `Bb`.

**Important:** When using wildcards, you don't use the `=` symbol; you use the **`LIKE`** keyword instead.

![](data/images/fingerprints.jpg)

### Example 4:

The below query only returns records from cities with names starting with the character `I` (or with the character `i`, since `LIKE` in SQLite is case insensitive). One thing to notice is that we didn't use the `*` syntax in it. That is because we only want to get one column in the output (the `city` column), not all columns.

In [11]:
SELECT city 
FROM crime_scene 
WHERE city LIKE 'I%';

+--------------+
| city         |
+--------------+
| Iowa City    |
+--------------+
| Indianapolis |
+--------------+

### Exercise 4

Modify the above query to make it retrieve the names of the cities whose names start with an S (instead of an I).

In [12]:
SELECT city 
FROM crime_scene 
WHERE city LIKE 'S%';

+------------------+
| city             |
+------------------+
| Stockton         |
+------------------+
| Shawnee Mission  |
+------------------+
| Saint Louis      |
+------------------+
| Sequel City      |
+------------------+
| Santa Rosa       |
+------------------+
| Sunnyvale        |
+------------------+
| Staten Island    |
+------------------+
| Sioux City       |
+------------------+
| Spartanburg      |
+------------------+
| Savannah         |
+------------------+
| Shreveport       |
+------------------+
| San Diego        |
+------------------+
| Salt Lake City   |
+------------------+
| San Diego        |
+------------------+
| San Diego        |
+------------------+
| San Jose         |
+------------------+
| Seattle          |
+------------------+
| Springfield      |
+------------------+
| Saint Petersburg |
+------------------+
| Seminole         |
+------------------+
| Sioux Falls      |
+------------------+
| Stockton         |
+------------------+
| San Diego        |
+------------------+
| San Diego        |
+------------------+
| Sequel City      |
+------------------+
| Springfield      |
+------------------+

### Exercise 5

You surely noticed that there were duplicate city names in the above queries, which is because one city can have more than one crime associated with it. If we wanted to get only the unique names (i.e., if we wanted to drop all the duplicate records), we would use **`SELECT DISTINCT`** instead of `SELECT`.

Rewrite the query from the previous exercise so that it shows each city once and only once.

In [15]:
SELECT DISTINCT city 
FROM crime_scene 
WHERE city LIKE 'S%';

+------------------+
| city             |
+------------------+
| Stockton         |
+------------------+
| Shawnee Mission  |
+------------------+
| Saint Louis      |
+------------------+
| Sequel City      |
+------------------+
| Santa Rosa       |
+------------------+
| Sunnyvale        |
+------------------+
| Staten Island    |
+------------------+
| Sioux City       |
+------------------+
| Spartanburg      |
+------------------+
| Savannah         |
+------------------+
| Shreveport       |
+------------------+
| San Diego        |
+------------------+
| Salt Lake City   |
+------------------+
| San Jose         |
+------------------+
| Seattle          |
+------------------+
| Springfield      |
+------------------+
| Saint Petersburg |
+------------------+
| Seminole         |
+------------------+
| Sioux Falls      |
+------------------+

### Exercise 6

Try your hand at writing queries with wildcards like `%a` and `%A_a` and then explore a few more if you like.

In [17]:
SELECT DISTINCT city 
FROM crime_scene 
WHERE city LIKE 'a%';

+-----------+
| city      |
+-----------+
| Atlanta   |
+-----------+
| Austin    |
+-----------+
| Arlington |
+-----------+
| Albany    |
+-----------+

## The shiftiness of the witnesses

We found an interesting lead in the output of Exercise 3:

> A security camera in the neighborhood captured two figures who may be witnesses. The first one lives at 513 Hamilton Dr. The second one was tentatively identified by one of the police officers as a man named George, who lives on Independence St.

They might be the witnesses of our murder case. Let's investigate a bit more.

### Exercise 7

Write a query that identifies the first witness based on the clue.


In [18]:
select *
from individual
where address like '%Hamilton%'

+-----+------------------+-----------+-----------------+-------------+----------+
| id  | name             | driver_id | address         | ssn         | income   |
+-----+------------------+-----------+-----------------+-------------+----------+
| 201 | Fidel Mancktelow | 865056986 | 513 Hamilton Dr | 109-20-3709 | 70497.05 |
+-----+------------------+-----------+-----------------+-------------+----------+

**Answer.**

### Exercise 8

Write a query to identify the second witness.

In [23]:
select *
from individual
where name like 'George%' and address like '%Independence%';

+-----+--------------+------------+---------------------+-------------+---------+
| id  | name         | driver_id  | address             | ssn         | income  |
+-----+--------------+------------+---------------------+-------------+---------+
| 328 | George Logue | 8473576861 | 223 Independence St | 104-40-3980 | 89659.9 |
+-----+--------------+------------+---------------------+-------------+---------+

In [ ]:
## The joining of the alibis

So far, we have been asking questions that can be answered by considering data from a single table alone. But what if we need to ask more complex questions that simultaneously require data from two different tables? There's a way to ask a database to merge two tables together and bring the result. It's by using the **`JOIN`** and **`ON`** keywords. The most common way to join tables is using their primary and foreign keys.

### Example 5

Let's say we wanted to find out the ages of the people in the database (as of Jan 1, 2019), such that the result table has the individual's `name` (which you can find in the `individual` table) and their `age` (which is present in the `drivers` table). Here is how to do that:

In [24]:
SELECT individual.name, drivers.age
FROM drivers 
JOIN individual ON drivers.id = individual.driver_id
LIMIT 10

+---------------------+-----+
| name                | age |
+---------------------+-----+
| Herve Lilie         | 28  |
+---------------------+-----+
| Jacqueline Colam    | 27  |
+---------------------+-----+
| Keenan Treker       | 49  |
+---------------------+-----+
| Kimbell Everill     | 35  |
+---------------------+-----+
| Abramo De Bernardis | 90  |
+---------------------+-----+
| Iggy Duferie        | 20  |
+---------------------+-----+
| Helli Bartoleyn     | 55  |
+---------------------+-----+
| Lacy Havik          | 70  |
+---------------------+-----+
| Winn Weldon         | 88  |
+---------------------+-----+
| Marlena Heggison    | 46  |
+---------------------+-----+

There are a bunch of interesting things you can see in the above query. First, we called the *table* name and then the *column* name, separating them with dots, like this: `individual.name`, `drivers.age`. This is useful to differentiate between columns that are from different tables but have the same names.

Secondly, we used `FROM` and then the name of one of the tables (`drivers` in this case) and then `JOIN` followed by the second table (`individual`). This makes it very explicit which tables we want to join.

Finally, we used `ON` to specify the condition to join on. This last bit is very important, because it tells the database how to find the records in one table that correspond to the records in the other one. Here, we tell it to match people by their driver's licenses  (the `drivers.id` column and the `individual.driver_id` column), so that if a record in the `drivers` table has a certain `id` and one person in the `individual` table has the same `driver_id`, they should be matched in the output.

### Exercise 9

Write a query that shows the names of the witnesses and their associated interview transcripts. We give you a portion of the code, but you will need to replace the `...`s with the correct answers:

~~~sql
%%sql
SELECT individual.name, interrogation.description
FROM ...
JOIN ... ON ...
WHERE individual.id = 201 OR individual.id = 328;
~~~

**Hint:** The **`OR`** keyword above retrieves all people with either `individual.id = 201` <u>or</u> `individual.id = 328` (the IDs we obtained in the previous exercises). If we use `AND` instead, the query will not give any results because no single person has two different `id`s simultaneously

In [25]:
SELECT individual.name, interrogation.description
FROM individual
JOIN interrogation ON individual.id = interrogation.individual_id
WHERE individual.id = 201 OR individual.id = 328;

+------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| name             | description                                                                                                                                                                                                                                                                                                                                                                                      |
+------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Fidel Mancktelow | I heard a scream and then saw someone run away. I thought it was just a robbery, but then I saw the pool of blood and freaked out. The guy that escaped accidentally left a card from the "Muscle Hassle" Gym. The ID member number was damaged but it starts with a 33. A minute later, I heard tyre screeches and saw a car pass by real fast, and the plate had "ZDI" or something like that. |
+------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| George Logue     | And, for the rest of the way home, I recited to them the various exploits and triumphs of Hercule Poirot.                                                                                                                                                                                                                                                                                        |
+------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

## Cards on the table 🃏

Now you know enough SQL to solve the mystery. Let's go get the murderer and head on to the big reveal!

### Exercise 10

#### 10.1

Here's Fidel Mancktelow's testimony:

> I heard a scream and then saw someone run away. I thought it was just a robbery, but then I saw the pool of blood and freaked out. The guy that escaped accidentally left a card from the "Muscle Hassle" Gym. The ID member number was damaged but it starts with a 33. A minute later, I heard tyre screeches and saw a car pass by real fast, and the plate had "ZDI" or something like that.

Write a query on the `gym_affiliated` table that retrieves all people with `id`s that start with `33-` and whose `status` is `Muscle Hassle`.

In [27]:
select *
from gym_affiliated
where id like '33%' and status = 'Muscle Hassle';

+-------------+---------------+------------------+------------+---------------+
| id          | individual_id | name             | start_date | status        |
+-------------+---------------+------------------+------------+---------------+
| 33-939-1315 | 647           | Krista Fingleton | 2017-12-23 | Muscle Hassle |
+-------------+---------------+------------------+------------+---------------+
| 33-207-7375 | 146           | Hercule Durram   | 2019-06-12 | Muscle Hassle |
+-------------+---------------+------------------+------------+---------------+
| 33-193-8015 | 981           | Tris MacVagh     | 2016-08-15 | Muscle Hassle |
+-------------+---------------+------------------+------------+---------------+
| 33-063-0778 | 45            | Bron Fontanet    | 2019-07-07 | Muscle Hassle |
+-------------+---------------+------------------+------------+---------------+

#### 10.2

Now take the `individual_id`s you've just found and find their license plate numbers.

**Hint:** You will need to join the `individual` and the `drivers` tables.

In [29]:
select *
from individual
join drivers on individual.driver_id = drivers.id
where individual.id = 647 or individual.id = 146 or individual.id = 981 or individual.id = 45;

+-----+------------------+------------+------------------------------+-------------+----------+------------+-----+--------+--------+------------+-----------+--------+-----------+------------+----------------+
| id  | name             | driver_id  | address                      | ssn         | income   | id         | age | gender | height | hair_color | eye_color | plate  | car_make  | car_model  | car_model_year |
+-----+------------------+------------+------------------------------+-------------+----------+------------+-----+--------+--------+------------+-----------+--------+-----------+------------+----------------+
| 45  | Bron Fontanet    | 1008583170 | 70279 Ilene Hill             | 844-32-7352 | 59955.93 | 1008583170 | 78  | Female | 49     | white      | black     | KSJ-33 | Chevrolet | S10 Blazer | 1992           |
+-----+------------------+------------+------------------------------+-------------+----------+------------+-----+--------+--------+------------+-----------+--------+-----------+------------+----------------+
| 146 | Hercule Durram   | 3736669879 | 019 Sherman Center           | 500-48-6973 | 42626.13 | 3736669879 | 58  | Female | 66     | red        | brown     | BFW-03 | Chevrolet | Express    | 2009           |
+-----+------------------+------------+------------------------------+-------------+----------+------------+-----+--------+--------+------------+-----------+--------+-----------+------------+----------------+
| 647 | Krista Fingleton | 659712261  | 865 Mayfield Court           | 774-14-6683 | 71007.68 | 659712261  | 18  | Male   | 48     | white      | brown     | ONZ-43 | Ford      | Aerostar   | 1997           |
+-----+------------------+------------+------------------------------+-------------+----------+------------+-----+--------+--------+------------+-----------+--------+-----------+------------+----------------+
| 981 | Tris MacVagh     | 9927362927 | 8706 Blue Bill Park Junction | 232-23-7588 | 45976.29 | 9927362927 | 66  | Male   | 50     | white      | green     | ZDI-13 | Maserati  | Spyder     | 1989           |
+-----+------------------+------------+------------------------------+-------------+----------+------------+-----+--------+--------+------------+-----------+--------+-----------+------------+----------------+

## A crooked twist

You may be thinking that you've solved the mystery... But did you?

Turns out, there's more to this case than meets the eye. The prime suspect has a surprising ace up his sleeve, which means that this cold case is...

![](data/images/to_be_continued.png)

(aka, wait for the *sequel*! 🥁)

## Takeaways & conclusion

In this case, we learned the differences between spreadsheets and databases. We also built a foundation of basic SQL commands to extract data from a database. Specifically we performed `SELECT...FROM` queries and learned the `WHERE`, `DISTINCT`, `LIKE` and `JOIN` keywords.

When working with large datasets, SQL is a powerful tool that can help us navigate and understand data in ways that Excel cannot. Sometimes, it can even serve as the first stage of an exploratory data analysis and can help us answer questions all by itself.

## Appendix: Basic SQL cheat sheet

**`SELECT`**

```SQL
- SELECT * FROM table_name -- Select all columns from a table
- SELECT column_name(s) FROM table_name -- Select some columns from a table
- SELECT DISTINCT column_name(s) FROM table_name -- Select only the different values
- SELECT column_name(s) FROM table_name -- Select data filtered with the WHERE clause
  WHERE condition
```

**Operators**

- `=` - Equal
- `LIKE` - Search pattern. Use `%` as a wildcard. E.g., `%o%` matches "o", "bob", "blob", etc.

## Attribution

The database used in this notebook has been adapted from NUKnightLab's `sql-mysteries` database, [MIT license](https://github.com/NUKnightLab/sql-mysteries/blob/master/LICENSE), https://github.com/NUKnightLab/sql-mysteries

"Fingerprints example", Public Domain, https://www.flickr.com/photos/publicdomainreview/24866294309

"President answering the red phone during cold war", Marco Verch, CC-BY, https://www.flickr.com/photos/149561324@N03/29283544077


"Investigation image", June 17th, 2021, [Pexels license](https://www.pexels.com/es-es/license/), https://www.pexels.com/es-es/foto/gente-mano-boligrafo-crimen-8369520/



